### ISSUE TITLE
apply_textfooler_create_dataset.py

### ISSUE THREAD

**Issue by meghasky** [meghasky/foss_lab63#1]
#this block only applies textfooler and then created combined dataset, further anlsyis in next block

import torch
import torch.nn as nn
import numpy as np
import json
from transformers import DistilBertTokenizer, DistilBertForSequenceClassification
from datasets import load_dataset, Dataset
from tqdm import tqdm
from itertools import product
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score
from textattack.attack_recipes import TextFoolerJin2019
from textattack.models.wrappers import HuggingFaceModelWrapper
import matplotlib.pyplot as plt

# Load and split SST-2 dataset
dataset = load_dataset("glue", "sst2")
train_data = dataset["train"].select(range(5000))  # Take only 5,000 samples

num_samples = len(train_data) // 2
clean_data = train_data.select(range(num_samples))
perturb_data = train_data.select(range(num_samples, len(train_data)))

# Save split datasets
clean_data.to_json("sst2_clean_split.json")
perturb_data.to_json("sst2_perturb_split.json")

# Load pre-trained tokenizer and model
tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')
classifier = DistilBertForSequenceClassification.from_pretrained('distilbert-base-uncased', num_labels=2).to(torch.device("cuda" if torch.cuda.is_available() else "cpu"))

# Wrap model for TextAttack
wrapped_model = HuggingFaceModelWrapper(classifier, tokenizer)

# Initialize TextFooler Attack
attack = TextFoolerJin2019.build(wrapped_model)

# Apply adversarial attack
perturbed_texts = []
for example in tqdm(perturb_data, desc="Applying TextFooler Attack"):
    input_text = example['sentence']
    ground_truth_label = example['label']
    try:
        adv_example = attack.attack(input_text, ground_truth_label)
        adv_text = adv_example.perturbed_text()
    except Exception as e:
        print(f"Error generating adversarial example: {e}")
        adv_text = input_text
    perturbed_texts.append({"sentence": adv_text, "label": ground_truth_label})

# Convert to Dataset format
perturbed_dataset = Dataset.from_list(perturbed_texts)
perturbed_dataset.to_json("sst2_textfooler_split.json")
